File: fib.js
/**
 * 
 * @param {number} n 
 */
// function fib(n) {
//   if (n <= 1) {
//     return n;
//   }
//   return fib(n - 1) + fib(n - 2);
// }

function fib(n) {
  let a = 0, b = 1, c = 0;
  let i = 2;
  for (i = 0; i < 9; i++) {
    c = a + b;
    a = b;
    b = c;
  }

  return c;
}

function fibClosure(n) {
  function _fib() {
    let a = 0, b = 1, c = 0;
    return function fib() {
      c = a + b;
      a = b;
      b = c;
      return c
    }
  }

  const f = _fib()
  let ret = 0;
  for (let i = 0; i < n; i++) {
    ret = f();
  }
  return ret;
}


let n = 9;
let sum = 0;

for (let i = 0; i < 10; i++) {
  let begin = process.hrtime();
  let result = fib(n);
  let end = process.hrtime(begin);
  console.log(i, ">", result, end[1]);
  sum += end[1]
}
console.log("Avg:", sum / 10);
console.log("====================");

sum = 0;
for (let i = 0; i < 10; i++) {
  let begin = process.hrtime();
  let result = fibClosure(n);
  let end = process.hrtime(begin);
  console.log(i, ">", result, end[1]);
  sum += end[1]
}
console.log("Avg:", sum / 10);

File: fib.py
from datetime import datetime


def fib(n):
  x1= 0
  x2 = 1
  x3 = 0
  for i in range(n):
    x3 = x1 + x2
    x1 = x2
    x2 = x3
  return x3


def fib_next():
  x1= 0
  x2 = 1
  def next():
    nonlocal x1
    nonlocal x2
    x3 = x1 + x2
    x1, x2 = x2, x3
    return x3
  return next


def fib_closure(n):
  x3 = 0
  f = fib_next()
  for i in range(n):
    x3 = f()
  return x3


if __name__ == "__main__":
  sum = 0
  for i in range(10):
    start = datetime.now()
    result = fib(20)
    print(i,"> ", result)
    end = datetime.now()
    eclipse = end - start
    sum = sum + eclipse.microseconds
    print(eclipse)

  print("Average", sum / 10)
  print("------------------")

  sum = 0
  for i in range(10):
    start = datetime.now()
    result = fib_closure(20)
    print(i,"> ", result)
    end = datetime.now()
    eclipse = end - start
    sum = sum + eclipse.microseconds
    print(eclipse)

  print("Average", sum / 10)